src/Client/Protocol/LspConnection.cs

```diff
@@ -0,0 +1,1011 @@
+            }
+
+            return headers;
+        }
+
+        /// <summary>
+        ///     The connection's message-dispatch loop.
+        /// </summary>
+        /// <returns>
+        ///     A <see cref="Task"/> representing the loop's activity.
+        /// </returns>
+        async Task DispatchLoop()
+        {
+            await Task.Yield();
+
+            Log.Information("Dispatch loop started.");
+
+            try
+            {
+                while (_incoming.TryTake(out ServerMessage message, -1, _cancellation))
+                {
+                    if (message.Id != null)
+                    {
+                        // Request.
+                        if (message.Method == "$/cancelRequest")
+                            CancelRequest(message);
+                        else
+                            DispatchRequest(message);
+                    }
+                    else
+                    {
+                        // Notification.
+                        DispatchNotification(message);
+                    }
+                }
+            }
+            catch (OperationCanceledException operationCanceled)
+            {
+                // Like tears in rain
+                if (operationCanceled.CancellationToken != _cancellation)
+                    throw; // time to die
+            }
+            finally
+            {
+                Log.Information("Dispatch loop terminated.");
+            }
+        }
+
+        /// <summary>
+        ///     Dispatch a request.
+        /// </summary>
+        /// <param name="requestMessage">
+        ///     The request message.
+        /// </param>
+        private void DispatchRequest(ServerMessage requestMessage)
+        {
+            if (requestMessage == null)
+                throw new ArgumentNullException(nameof(requestMessage));
+
+            string requestId = requestMessage.Id.ToString();
+            Log.Verbose("Dispatching incoming {RequestMethod} request {RequestId}...", requestMessage.Method, requestId);
+
+            CancellationTokenSource requestCancellation = CancellationTokenSource.CreateLinkedTokenSource(_cancellation);
+            _requestCancellations.TryAdd(requestId, requestCancellation);
+
+            Task<object> handlerTask = _dispatcher.TryHandleRequest(requestMessage.Method, requestMessage.Params, requestCancellation.Token);
+            if (handlerTask == null)
+            {
+                Log.Warning("Unable to dispatch incoming {RequestMethod} request {RequestId} (no handler registered).", requestMessage.Method, requestId);
+
+                _outgoing.TryAdd(
+                    new JsonRpcMessages.MethodNotFound(requestMessage.Id, requestMessage.Method)
+                );
+            }
+
+#pragma warning disable CS4014 // Continuation does the work we need; no need to await it as this would tie up the dispatch loop.
+            handlerTask.ContinueWith(_ =>
```

Comment:
This will throw if `handlerTask` is null, should probably `return` in the above if.

Reply:
Hah! Yes, good catch :)